syntax = "proto3";
package saber;

enum SessionState {
  SS_CONNECTED = 0;
  SS_DISCONNECTED = 1;
  SS_EXPIRED = 2;
  SS_AUTHFAILED = 3;
}

enum EventType {
  ET_NONE = 0;
  ET_NODE_CREATED = 1;
  ET_NODE_DELETED = 2;
  ET_NODE_DATA_CHANGED = 3;
  ET_NODE_CHILDREN_CHANGED = 4;
  ET_DATA_WATCH_REMOVED = 5;
  ET_CHILD_WATCH_REMOVED = 6;
}

message WatchedEvent {
  SessionState state = 1;
  EventType type = 2;
  bytes path = 3;
}

enum NodeType { NT_PERSISTENT = 0; }

enum ResponseCode {
  RC_OK = 0;
  RC_FAILED = 1;
  RC_NO_NODE = 2;
  RC_NO_PARENT = 3;
  RC_NODE_EXISTS = 4;
  RC_CHILDREN_EXISTS = 5;
  RC_BAD_VERSION = 6;
  RC_NO_AUTH = 7;
  RC_UNKNOWN = 8;
}

message Stat {
  uint32 group_id = 1;
  uint64 created_id = 2;
  uint64 modified_id = 3;
  uint64 created_time = 4;
  uint64 modified_time = 5;
  int32 version = 6;
  int32 children_version = 7;
  int32 acl_version = 8;
  uint32 data_len = 9;
  uint32 children_num = 10;
  uint64 children_id = 11;
}

message Id {
  bytes scheme = 1;
  bytes id = 2;
}

message ACL {
  int32 perms = 1;
  Id id = 2;
}

message ConnectRequest {
  uint64 session_id = 1;
  uint64 timeout = 2;
}

message ConnectResponse {
  ResponseCode code = 1;
  uint64 session_id = 2;
  uint64 timeout = 3;
}

message CreateRequest {
  bytes path = 1;
  bytes data = 2;
  repeated ACL acl = 3;
  NodeType type = 4;
}

message CreateResponse {
  ResponseCode code = 1;
  bytes path = 2;
}

message DeleteRequest {
  bytes path = 1;
  int32 version = 2;
}

message DeleteResponse { ResponseCode code = 1; }

message ExistsRequest {
  bytes path = 1;
  bool watch = 2;
}

message ExistsResponse {
  ResponseCode code = 1;
  Stat stat = 2;
}

message GetDataRequest {
  bytes path = 1;
  bool watch = 2;
}

message GetDataResponse {
  ResponseCode code = 1;
  bytes data = 2;
  Stat stat = 3;
}

message SetDataRequest {
  bytes path = 1;
  bytes data = 2;
  int32 version = 3;
}

message SetDataResponse {
  ResponseCode code = 1;
  Stat stat = 2;
}

message GetACLRequest { bytes path = 1; }

message GetACLResponse {
  ResponseCode code = 1;
  repeated ACL acl = 2;
  Stat stat = 3;
}

message SetACLRequest {
  bytes path = 1;
  repeated ACL acl = 2;
  int32 version = 3;
}

message SetACLResponse {
  ResponseCode code = 1;
  Stat stat = 2;
}

message GetChildrenRequest {
  bytes path = 1;
  bool watch = 2;
}

message GetChildrenResponse {
  ResponseCode code = 1;
  Stat stat = 2;
  repeated bytes children = 3;
}

message Master {
  bytes host = 1;
  int32 port = 2;
}

enum MessageType {
  MT_NOTIFICATION = 0;
  MT_CREATE = 1;
  MT_DELETE = 2;
  MT_EXISTS = 3;
  MT_GETDATA = 4;
  MT_SETDATA = 5;
  MT_GETACL = 6;
  MT_SETACL = 7;
  MT_GETCHILDREN = 8;
  MT_MASTER = 9;
  MT_PING = 10;
  MT_CONNECT = 11;
}

message SaberMessage {
  MessageType type = 1;
  bytes data = 2;
  bytes extra_data = 3;
}

message DataNode {
  Stat stat = 1;
  bytes name = 2;
  bytes data = 3;
  repeated ACL acl = 4;
  repeated bytes children = 5;
}
